slo_generator/backends/cloud_monitoring_mql.py

```diff
@@ -0,0 +1,262 @@
+
+        Returns:
+            float: SLI value.
+        """
+        measurement: dict = slo_config['spec']['service_level_indicator']
+        query: str = measurement['query']
+        series: list[TimeSeriesData] = self.query(query=query, window=window)
+        sli_value: float = series[0].point_data[0].values[0].double_value
+        LOGGER.debug(f"SLI value: {sli_value}")
+        return sli_value
+
+    def query(self, query: str, window: int) -> list[TimeSeriesData]:
+        """Query timeseries from Cloud Monitoring using MQL.
+
+        Args:
+            query (str): MQL query.
+            window (int): Window size (in seconds).
+
+        Returns:
+            list: List of timeseries objects.
+        """
+        # Enrich query to aggregate and reduce the time series over the
+        # desired window.
+        formatted_query: str = self._fmt_query(query, window)
+        request = metric_service.QueryTimeSeriesRequest({
+            'name': self.parent,
+            'query': formatted_query
+        })
+        timeseries_pager: QueryTimeSeriesPager = self.client.query_time_series(
+            request)
+        timeseries: list = list(timeseries_pager)  # convert pager to flat list
+        LOGGER.debug(pprint.pformat(timeseries))
+        return timeseries
+
+    @staticmethod
+    def count(timeseries: list[TimeSeriesData]) -> int:
+        """Count events in time series assuming it was aligned with ALIGN_SUM
+        and reduced with REDUCE_SUM (default).
+
+        Args:
+            :obj:`monitoring_v3.TimeSeries`: Timeseries object.
+
+        Returns:
+            int: Event count.
+        """
+        try:
+            return timeseries[0].point_data[0].values[0].int64_value
+        except (IndexError, AttributeError) as exception:
+            LOGGER.debug(exception, exc_info=True)
+            return NO_DATA  # no events in timeseries
+
+    @staticmethod
+    def _fmt_query(query: str, window: int) -> str:
+        """Format MQL query:
+
+        * If the MQL expression has a `window` placeholder, replace it by the
+        current window. Otherwise, append it to the expression.
+
+        * If the MQL expression has a `every` placeholder, replace it by the
+        current window. Otherwise, append it to the expression.
+
+        * If the MQL expression has a `group_by` placeholder, replace it.
+        Otherwise, append it to the expression.
+
+        Args:
+            query (str): Original query in YAMLconfig.
+            window (int): Query window (in seconds).
+
+        Returns:
+            str: Formatted query.
+        """
+        formatted_query: str = query.strip()
+        if 'group_by' in formatted_query:
+            formatted_query = re.sub(r'\|\s+group_by\s+\[.*\]\s*',
+                                     '| group_by [] ', formatted_query)
+        else:
+            formatted_query += '| group_by [] '
+        for mql_time_interval_keyword in ['within', 'every']:
```

Comment:
This code can be hard to read and maintain. Perhaps I should add some practical examples of inputs and outputs. For example the queries I used in the unit tests.